pkg/syncer/resourcesync/controller.go

```diff
@@ -365,79 +380,128 @@ func (c *Controller) checkSSAR(ctx context.Context, gvr schema.GroupVersionResou
+		if apiResource.Name == gvr.Resource {
+			var resourceScope apiextensionsv1.ResourceScope
+			if apiResource.Namespaced {
+				resourceScope = apiextensionsv1.NamespaceScoped
+			} else {
+				resourceScope = apiextensionsv1.ClusterScoped
+			}
+
+			return &informer.GVRPartialMetadata{
+					Names: apiextensionsv1.CustomResourceDefinitionNames{
+						Kind:     apiResource.Kind,
+						Singular: apiResource.SingularName,
+					},
+					Scope: resourceScope,
+				},
+				nil
 		}
 	}
+	return nil, fmt.Errorf("unable to retrieve discovery for GVR: %s", gvr)
 }
 
-func (c *Controller) startSyncerInformer(ctx context.Context, gvr schema.GroupVersionResource, syncTarget *workloadv1alpha1.SyncTarget) {
+func (c *controller) notifySubscribers(ctx context.Context) {
 	logger := klog.FromContext(ctx)
 
-	c.mutex.Lock()
-	defer c.mutex.Unlock()
+	c.subscribersLock.Lock()
+	defer c.subscribersLock.Unlock()
 
-	if _, ok := c.syncerInformerMap[gvr]; ok {
-		logger.V(2).Info("Informer is started already")
-		return
+	for index, ch := range c.subscribers {
+		logger.V(4).Info("Attempting to notify subscribers", "index", index)
+		select {
+		case ch <- struct{}{}:
+			logger.V(4).Info("Successfully notified subscriber", "index", index)
+		default:
+			logger.V(4).Info("Unable to notify subscriber - channel full", "index", index)
+		}
 	}
+}
 
-	syncTargetKey := workloadv1alpha1.ToSyncTargetKey(c.syncTargetClusterName, c.syncTargetName)
+func (c *controller) addGVR(ctx context.Context, gvr schema.GroupVersionResource) error {
+	logger := klog.FromContext(ctx)
 
-	upstreamInformer := kcpdynamicinformer.NewFilteredDynamicInformer(c.upstreamDynamicClusterClient, gvr, resyncPeriod, cache.Indexers{
-		kcpcache.ClusterIndexName:             kcpcache.ClusterIndexFunc,
-		kcpcache.ClusterAndNamespaceIndexName: kcpcache.ClusterAndNamespaceIndexFunc}, func(o *metav1.ListOptions) {},
-	)
-	downstreamInformer := dynamicinformer.NewFilteredDynamicInformer(c.downstreamDynamicClient, gvr, metav1.NamespaceAll, resyncPeriod, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc}, func(o *metav1.ListOptions) {
-		o.LabelSelector = workloadv1alpha1.InternalDownstreamClusterLabel + "=" + syncTargetKey
-	})
+	c.gvrsToWatchLock.Lock()
+	defer c.gvrsToWatchLock.Unlock()
 
-	for _, handler := range c.upstreamEventHandlers {
-		upstreamInformer.Informer().AddEventHandler(handler(gvr))
+	if _, ok := c.gvrsToWatch[gvr]; ok {
+		logger.V(2).Info("Informer is started already")
+		return nil
 	}
 
-	for _, handler := range c.downstreamEventHandlers {
-		downstreamInformer.Informer().AddEventHandler(handler(gvr))
+	partialMetadata, err := c.getGVRPartialMetadata(gvr)
+	if err != nil {
+		return err
 	}
 
-	logger.V(2).Info("Start informer for gvr")
-	syncerCtx, cancel := context.WithCancel(ctx)
+	c.gvrsToWatch[gvr] = *partialMetadata
 
-	go downstreamInformer.Informer().Run(syncerCtx.Done())
-	go upstreamInformer.Informer().Run(syncerCtx.Done())
+	c.notifySubscribers(ctx)
```

Comment:
don't do this within the lock

Reply:
Fixed in commit da89d306186c852a5f9f9ab2a3a549f896f5f7e0